L1Trigger/TrackTrigger/plugins/TTStubBuilder.cc

```diff
@@ -0,0 +1,327 @@
+template< >
+void TTStubBuilder< Ref_Phase2TrackerDigi_ >::produce( edm::Event& iEvent, const edm::EventSetup& iSetup )
+{
+  //Retrieve tracker topology from geometry                                                                                                              
+  edm::ESHandle<TrackerTopology> tTopoHandle;
+  iSetup.get<TrackerTopologyRcd>().get(tTopoHandle);
+  const TrackerTopology* const tTopo = tTopoHandle.product();
+  edm::ESHandle< TrackerGeometry > tGeomHandle;
+  iSetup.get< TrackerDigiGeometryRecord >().get( tGeomHandle );
+  const TrackerGeometry* const theTrackerGeom = tGeomHandle.product();
+	
+  /// Prepare output
+  std::auto_ptr< edmNew::DetSetVector< TTCluster< Ref_Phase2TrackerDigi_ > > > TTClusterDSVForOutput( new edmNew::DetSetVector< TTCluster< Ref_Phase2TrackerDigi_ > > );
+  std::auto_ptr< edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > > TTStubDSVForOutputTemp( new edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > );
+  std::auto_ptr< edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > > TTStubDSVForOutputAccepted( new edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > );
+  std::auto_ptr< edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > > TTStubDSVForOutputRejected( new edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > > );
+
+  /// Get the Clusters already stored away
+  edm::Handle< edmNew::DetSetVector< TTCluster< Ref_Phase2TrackerDigi_ > > > clusterHandle;
+  iEvent.getByToken( clustersToken, clusterHandle );
+
+  /// Get the maximum number of stubs per ROC
+  /// (CBC3-style)
+  //  unsigned maxStubs = theStackedTracker->getCBC3MaxStubs();
+  unsigned maxStubs = 3;
+
+  for (auto gd=theTrackerGeom->dets().begin(); gd != theTrackerGeom->dets().end(); gd++) {
+      DetId detid = (*gd)->geographicalId();
+      if(detid.subdetId()!=StripSubdetector::TOB && detid.subdetId()!=StripSubdetector::TID ) continue; // only run on OT
+      if(!tTopo->isLower(detid) ) continue; // loop on the stacks: choose the lower arbitrarily
+      DetId lowerDetid = detid;
+      DetId upperDetid = tTopo->partnerDetId(detid);
+      DetId stackDetid = tTopo->stack(detid);
+
+    /// Go on only if both detectors have Clusters
+    if ( clusterHandle->find( lowerDetid ) == clusterHandle->end() ||
+         clusterHandle->find( upperDetid ) == clusterHandle->end() )
+      continue;
+
+    /// Get the DetSets of the Clusters
+    edmNew::DetSet< TTCluster< Ref_Phase2TrackerDigi_ > > lowerClusters = (*clusterHandle)[ lowerDetid ];
+    edmNew::DetSet< TTCluster< Ref_Phase2TrackerDigi_ > > upperClusters = (*clusterHandle)[ upperDetid ];
+
+    /// If there are Clusters in both sensors
+    /// you can try and make a Stub
+    /// This is ~redundant
+    if ( lowerClusters.size() == 0 || upperClusters.size() == 0 )
+      continue;
+
+    /// Create the vectors of objects to be passed to the FastFillers
+    std::vector< TTCluster< Ref_Phase2TrackerDigi_ > > tempInner; 
```

Comment:
General comment: what do you mean with inner/outer? In which system of reference you would like to set this definition?